stl/inc/memory

```diff
@@ -2416,26 +2437,298 @@ bool atomic_compare_exchange_weak_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
 template <class _Ty>
-bool atomic_compare_exchange_strong_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value,
-    memory_order, memory_order) { // atomically compare and exchange
+_CXX20_DEPRECATE_OLD_SHARED_PTR_ATOMIC_SUPPORT bool atomic_compare_exchange_strong_explicit(shared_ptr<_Ty>* _Ptr,
+    shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value, memory_order, memory_order) { // atomically compare and exchange
     return _STD atomic_compare_exchange_weak_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
 template <class _Ty>
-bool atomic_compare_exchange_strong(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
+_CXX20_DEPRECATE_OLD_SHARED_PTR_ATOMIC_SUPPORT bool atomic_compare_exchange_strong(
+    shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
     // atomically compare and exchange
     return _STD atomic_compare_exchange_strong_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
+#if _HAS_CXX20
+template <class _Ty>
+struct alignas(2 * sizeof(void*)) _Atomic_ptr_base {
+protected:
+    constexpr _Atomic_ptr_base() noexcept = default;
+
+    _Atomic_ptr_base(_Ty* const _Px, _Ref_count_base* const _Ref) noexcept
+        : _Ptr(_Px), _Refptr(reinterpret_cast<const uintptr_t>(_Ref)) {}
+
+    _Ref_count_base* _LockAndLoad() const noexcept {
+        uintptr_t _Ref = _Refptr & ~1;
+        while (!_Refptr.compare_exchange_strong(_Ref, _Ref | 1)) {
+            _Ref &= ~1;
+            _YIELD_PROCESSOR();
+        }
+        return reinterpret_cast<_Ref_count_base*>(_Ref);
+    }
+
+    _Ref_count_base* _LockAndExchange(_Ref_count_base* const _Value) const noexcept {
+        uintptr_t _UValue = reinterpret_cast<uintptr_t>(_Value);
+        _UValue |= 1;
+        uintptr_t _Ref = _Refptr & ~1;
+        while (!_Refptr.compare_exchange_strong(_Ref, _UValue)) {
+            _Ref &= ~1;
+            _YIELD_PROCESSOR();
+        }
+        return reinterpret_cast<_Ref_count_base*>(_Ref);
+    }
+
+    void _Unlock() const noexcept {
+        _Refptr &= ~1;
+    }
+
+    void _StoreAndUnlock(_Ref_count_base* const _Value) const noexcept {
+        _Refptr = reinterpret_cast<uintptr_t>(_Value);
+    }
+
+    _Ty* _Ptr                         = nullptr;
+    mutable atomic<uintptr_t> _Refptr = 0;
+};
+
+template <class _Ty>
+struct atomic<shared_ptr<_Ty>> : _Atomic_ptr_base<_Ty> {
+private:
+    using _Base = _Atomic_ptr_base<_Ty>;
+
+public:
+    using value_type = shared_ptr<_Ty>;
+
+    static constexpr bool is_always_lock_free = false;
+
+    _NODISCARD bool is_lock_free() const noexcept {
+        return false;
+    }
+
+    void store(const shared_ptr<_Ty> _Value, const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_store_memory_order(_Order);

```

Comment:
This can be more efficient as:

```
const auto _Old_ref = this->_Lock_and_load();
std::swap(this->_Ptr, _Value._Ptr);
this->_Rep.store(_Value._Ref); // also unlocks
_Value._Ref = _Old_ref;
```

Reply:
Done